### ISSUE TITLE
Test t/43_fts3.t extremely slow on DBD-SQLite-1.71_05 built with -DDEBUGGING

### ISSUE THREAD

**Issue by andk** [DBD-SQLite/DBD-SQLite#98]
The good news is that all tests pass. A pain point with -DDEBUGGING is the slowness of the test t/43_fts3.t. This test has built in four note() diagnostics lines. The slowness seems to be independent of the perl version. Here are the four diagnostic outputs I just observed:

``` text
# long doc (487225 chars): insert in 0.3018 secs, select in 0.0010 secs
# long doc (487225 chars): insert in 0.2968 secs, select in 0.0019 secs
# long doc (487225 chars): insert in 327.7886 secs, select in 0.0006 secs
# long doc (487225 chars): insert in 334.0167 secs, select in 0.0009 secs
```

**Comment by charsbar** [comment]
Thanks. This slowdown starts from 1.67_04, so this must be something from the strict unicode mode introduced then, or some of the unicode related assertion in perl as it only happens when perl is built with -DDEBUGGING.

@FGasper, do you have any idea?

**Comment by FGasper** [comment]
Try Devel::NYTProf?

-F

**Comment by FGasper** [comment]
@andk Does this make the test acceptably faster?
```
diff --git a/t/43_fts3.t b/t/43_fts3.t
index 32340c1..6b9d306 100644
--- a/t/43_fts3.t
+++ b/t/43_fts3.t
@@ -59,7 +59,7 @@ sub Unicode_Word_tokenizer { # see also: Search::Tokenizer

 use DBD::SQLite;

-for my $string_mode (DBD_SQLITE_STRING_MODE_BYTES, DBD_SQLITE_STRING_MODE_UNICODE_STRICT) {
+for my $string_mode (DBD_SQLITE_STRING_MODE_BYTES, DBD_SQLITE_STRING_MODE_UNICODE_NAIVE) {

   # connect
   my $dbh = connect_ok( RaiseError => 1, sqlite_string_mode => $string_mode );
@@ -106,7 +106,7 @@ for my $string_mode (DBD_SQLITE_STRING_MODE_BYTES, DBD_SQLITE_STRING_MODE_UNICOD
     # the 'offsets' function should return integer offsets for the words in the MATCH query
     my $sql_offsets = "SELECT offsets(try_$fts) FROM try_$fts WHERE content MATCH ?";
     $result = $dbh->selectcol_arrayref($sql_offsets, undef, 'une');
-    my $offset_une = $string_mode == DBD_SQLITE_STRING_MODE_UNICODE_STRICT ? $ix_une_utf8 : $ix_une_native;
+    my $offset_une = $string_mode == DBD_SQLITE_STRING_MODE_UNICODE_NAIVE ? $ix_une_utf8 : $ix_une_native;
     my $expected_offsets = "0 0 $offset_une 3";
     is_deeply($result, [$expected_offsets], "offsets ($fts, string_mode=$string_mode)");
```
It _may_ be that Perl’s UTF-8 validation/decoding is just that much slower under -DDEBUGGING. In that case, maybe the test should detect and favour the “naive” mode for debugging builds.

**Comment by andk** [comment]
No. I'm measuring on a faster machine today, so I present you the four lines before the patch and after the patch:
``` text
# long doc (487225 chars): insert in 0.0746 secs, select in 0.0006 secs
# long doc (487225 chars): insert in 0.0742 secs, select in 0.0003 secs
# long doc (487225 chars): insert in 107.1819 secs, select in 0.0009 secs
# long doc (487225 chars): insert in 108.4512 secs, select in 0.0013 secs

# long doc (487225 chars): insert in 0.0746 secs, select in 0.0006 secs
# long doc (487225 chars): insert in 0.0734 secs, select in 0.0003 secs
# long doc (487225 chars): insert in 107.7872 secs, select in 0.0004 secs
# long doc (487225 chars): insert in 108.6713 secs, select in 0.0009 secs

```

**Comment by FGasper** [comment]
![image](https://user-images.githubusercontent.com/844867/156082561-161ccb8b-6ad2-4243-ab86-3170dbb88048.png)
Curious … Devel::NYTProf seems to think the slowness is in an area I didn’t touch.

**Comment by Leont** [comment]
A `-DDEBUGGING` build automatically enables debugging mode for the utf8 cache. If this is a problem, you can set `${^UTF8CACHE} = 1` to disable that debugging code.

**Comment by FGasper** [comment]
@leont Thank you!

@andk This resolves the slowdown for me:
```
diff --git a/t/43_fts3.t b/t/43_fts3.t
index b57eb4a..9897323 100644
--- a/t/43_fts3.t
+++ b/t/43_fts3.t
@@ -7,6 +7,8 @@ use Test::More;
 use if -d ".git", "Test::FailWarnings";
 use DBD::SQLite;

+# Avoid slowdown with -DDEBUGGING:
+${^UTF8CACHE} = 1;

 my @texts = ("il <E9>tait une berg<E8>re",
              "qui gardait ses moutons",
```

**Comment by andk** [comment]
Confirmed. Thanks!

**Comment by charsbar** [comment]
Thank you everyone. Shipped 1.71_06.
